L1Trigger/L1THGCalUtilities/plugins/ntuples/HGCalTriggerNtupleHGCConcentratorData.cc

```diff
@@ -0,0 +1,155 @@
+  tree.Branch(withPrefix("subdet"), &econ_subdet_);
+  tree.Branch(withPrefix("zside"), &econ_side_);
+  tree.Branch(withPrefix("layer"), &econ_layer_);
+  tree.Branch(withPrefix("wafer"), &econ_wafer_);
+  tree.Branch(withPrefix("waferu"), &econ_waferu_);
+  tree.Branch(withPrefix("waferv"), &econ_waferv_);
+  tree.Branch(withPrefix("wafertype"), &econ_wafertype_);
+  tree.Branch(withPrefix("panel_number"), &econ_panel_number_);
+  tree.Branch(withPrefix("panel_sector"), &econ_panel_sector_);
+  tree.Branch(withPrefix("index"), &econ_index_);
+  tree.Branch(withPrefix("data"), &econ_data_);
+}
+
+void HGCalTriggerNtupleHGCConcentratorData::fill(const edm::Event& e, const edm::EventSetup& es) {
+  // retrieve trigger cells
+  edm::Handle<l1t::HGCalConcentratorDataBxCollection> concentrator_data_h;
+  e.getByToken(concentrator_data_token_, concentrator_data_h);
+  const l1t::HGCalConcentratorDataBxCollection& concentrator_data = *concentrator_data_h;
+
+  // retrieve geometry
+  es.get<CaloGeometryRecord>().get(geometry_);
+
+  triggerTools_.eventSetup(es);
+
+  clear();
+  for (auto econ_itr = concentrator_data.begin(0); econ_itr != concentrator_data.end(0); econ_itr++) {
+    econ_n_++;
+    // hardware data
+    DetId panelId(econ_itr->detId());
+    int panel_sector = -999;
+    int panel_number = -999;
+    if (panelId.det() == DetId::Forward) {
+      HGCalDetId panelIdHGCal(panelId);
+      if (panelId.subdetId() == ForwardSubdetector::HGCHEB) {
+        panel_number = panelIdHGCal.wafer();
+      } else {
+        panel_sector = (panelIdHGCal.wafer() >> kSectorOffset_) & kSectorMask_;
+        panel_number = (panelIdHGCal.wafer() >> kPanelOffset_) & kPanelMask_;
+      }
+    } else if (panelId.det() == DetId::HGCalHSc) {
+      HGCScintillatorDetId panelIdSci(panelId);
+      panel_sector = panelIdSci.iphi();
+      panel_number = panelIdSci.ietaAbs();
+    }
+    econ_panel_number_.emplace_back(panel_number);
+    econ_panel_sector_.emplace_back(panel_sector);
+    econ_id_.emplace_back(econ_itr->detId());
+    econ_side_.emplace_back(triggerTools_.zside(panelId));
+    econ_layer_.emplace_back(triggerTools_.layerWithOffset(panelId));
+    // V9 detids
+    if (panelId.det() == DetId::HGCalTrigger) {
+      HGCalTriggerDetId idv9(panelId);
+      econ_subdet_.emplace_back(idv9.subdet());
+      econ_waferu_.emplace_back(idv9.waferU());
+      econ_waferv_.emplace_back(idv9.waferV());
+      econ_wafertype_.emplace_back(idv9.type());
+    } else if (panelId.det() == DetId::HGCalHSc) {
+      HGCScintillatorDetId idv9(panelId);
+      econ_subdet_.emplace_back(idv9.subdet());
+      econ_waferu_.emplace_back(-999);
+      econ_waferv_.emplace_back(-999);
+      econ_wafertype_.emplace_back(idv9.type());
+    }
+    // V8 detids
+    else {
+      HGCalDetId idv8(panelId);
+      econ_subdet_.emplace_back(panelId.subdetId());
+      econ_wafer_.emplace_back(idv8.wafer());
+      econ_wafertype_.emplace_back(idv8.waferType());
+    }
```

Comment:
This shouldn't happen in principle given that the `HGCalConcentratorData` ids are necessarily `HGCalTriggerDetId`